### ISSUE TITLE
Silent authentication mechanism failures

### ISSUE THREAD

**Issue by bender2k14** [jstedfast/MailKit#761]
The mail clients ([IMAP](https://github.com/jstedfast/MailKit/blob/master/MailKit/Net/Imap/ImapClient.cs#L592), [SMTP](https://github.com/jstedfast/MailKit/blob/master/MailKit/Net/Smtp/SmtpClient.cs#L241), and [POP3](https://github.com/jstedfast/MailKit/blob/master/MailKit/Net/Pop3/Pop3Client.cs#L340)) allow multiple authentication mechanisms to be considered when instructed to authenticate.  My experience with this is that if one mechanism fails, then another is tried.  If all mechanisms fail, then some authentication exception is thrown.

Recently I was using the `SmptClient` to connect to a server with some odd behavior.  Consider these two protocol logs.

     1 Connected to smtp://mail.*********.com:587/?starttls=when-available
     2 S: 220 *********.com ESMTP
     3 C: EHLO [192.168.0.4]
     4 S: 250-*********.com
     5 S: 250-STARTTLS
     6 S: 250-PIPELINING
     7 S: 250-8BITMIME
     8 S: 250-SIZE 0
     9 S: 250 AUTH LOGIN PLAIN CRAM-MD5
    10 C: STARTTLS
    11 S: 220 ready for tls
    12 C: EHLO [192.168.0.4]
    13 S: 250-*********.com
    14 S: 250-PIPELINING
    15 S: 250-8BITMIME
    16 S: 250-SIZE 0
    17 S: 250 AUTH LOGIN PLAIN CRAM-MD5
    18 C: AUTH CRAM-MD5
    19 S: 334 ****************************************************
    20 C: ************************************************************************
    21 S: 535 authentication failed (#5.7.1)
    22 C: AUTH PLAIN ********************************************
    23 S: 235 ok, go ahead (#2.0.0)
    24 C: QUIT
    25 S: 221 *********.com
    26 
    27 Connected to smtp://mail.*********.com:587/?starttls=when-available
    28 S: 220 *********.com ESMTP
    29 C: EHLO [192.168.0.4]
    30 S: 250-*********.com
    31 S: 250-STARTTLS
    32 S: 250-PIPELINING
    33 S: 250-8BITMIME
    34 S: 250-SIZE 0
    35 S: 250 AUTH LOGIN PLAIN CRAM-MD5
    36 C: STARTTLS
    37 S: 220 ready for tls
    38 C: EHLO [192.168.0.4]
    39 S: 250-*********.com
    40 S: 250-PIPELINING
    41 S: 250-8BITMIME
    42 S: 250-SIZE 0
    43 S: 250 AUTH LOGIN PLAIN CRAM-MD5
    44 C: AUTH CRAM-MD5
    45 S: 334 ****************************************************
    46 C: ************************************************************************
    47 S: 235 ok, go ahead (#2.0.0)
    48 C: QUIT
    49 S: 221 *********.com

The redacted parts were identical in both cases.

As we can see from the protocol logs (on lines 9, 17, 35, and 43), this server claims to allow the `CRAM-MD5` authentication mechanism.  However, the administrator of this server informed me that `CRAM-MD5` is actually not supported by this server and that `CRAM-MD5` only appears in the list of supported mechanisms for backwards compatibility with legacy systems.  Sometimes (as in the second protocol log above), it appears that authentication via `CRAM-MD5` was successful.  Actually what happened is that the server ignored the provided data and instead fell back to IMAP before SMTP (which is similar to [POP before SMTP](https://en.wikipedia.org/wiki/POP_before_SMTP)) and approved the authentication because of a recent successful IMAP authentication.  At other times (as in the first protocol log), no recent successful IMAP authentication exists, so the authentication via `CRAM-MD5` appears to fail.

Up to this point, everything seems to work in spite of this server's treatment of `CRAM-MD5`.  However, each time that an authentication attempt via `CRAM-MD5` failed (i.e. the fallbacks to POP before SMTP and IMAP before SMTP failed), this fact was recorded in their logs.  Then a separate application ([Fail2ban](https://www.fail2ban.org/wiki/index.php/Main_Page)) scans their logs looking for failed authentication attempts.  In the meantime, I increased the testing of my application by running more and more instances in parallel (behind the same public IP address).  Eventually I had racked up enough failed `CRAM-MD5` authentication attempts in a short enough time span that I hit their temporary ban threshold for my public IP address.  Then all the instances of my application reported an inability to make contact with the SMTP server.

I fixed this issue by removing `CRAM-MD5` from the set of authentication mechanisms that MailKit considers.  I was hesitant to do this at first since I want my application to work with more than just this one particular SMTP server.  If I removed an authentication mechanism each time I found a server that had an issue with it, then it wouldn't be long before wouldn't have any mechanisms left.  However, it seems that [`CRAM-MD5` is deprecated](https://www.gnu.org/software/gsasl/manual/html_node/CRAM_002dMD5.html), so I decided to remove it.

I am sharing this here in the hope that it helps other MailKit users and maybe also helps to improve MailKit.

The first improvement that comes to mind is to change MailKit to remove `CRAM-MD5` from the default set of SMTP authentication mechanisms.  This seems pretty reasonable if it really is the case that this mechanism is deprecated.

The second improvement I thought of is an easier way to know that some authentication mechanism failed.  My application suffered from this behavior for months before I finally figured out the problem.  Granted, most of that time was spent implementing other features and fixing other bugs.  I would have figured it out sooner if I would have worked on nothing else.  The last breakthrough came when I was lucky enough to capture the first protocol log above and compare it with my application logs, which reported no exceptional behavior.  Given that this is MailKit's GitHub, it may seem obvious to begin by looking closely at the protocol log.  However, when this issue first began, I had no idea what the problem was.  I was checking networking cables, routers, `ping`ing the server, and `tracerout`ing the server.  It would have been really helpful if MailKit would have notified me about the failed `CRAM-MD5` authentication attempts.  Then I would have been able to spend almost no time debugging before I had a solid lead to investigate further.

Can MailKit expose some visibility into failed authentication attempts per mechanism?  Maybe this feature already exists and I am just unaware of it.

**Comment by bender2k14** [comment]
Yes, I linked to that property above.  And yes, my fix was to explicitly remove `CRAM-MD5` from that set.

What do you mean by "to even try by using"?  Do you mean "will be tried by reading"?

**Comment by jstedfast** [comment]
No, what I meant was that you can check what mechanisms are supported there so that you know which MailKit.Security.SaslMechanism's you can try using with any given server.

So for example, if you know `PLAIN` is supported, you can do this:

```csharp
var sasl = SaslMechanism.Create ("PLAIN", uri, credentials);
client.Authenticate (sasl);
```

**Comment by jstedfast** [comment]
The other option is what you mentioned which is to do:

```csharp
client.AuthenticationMechanisms.Remove ("CRAM-MD5");
client.Authenticate (...);
```

**Comment by bender2k14** [comment]
> ...you can check what mechanisms are supported there...

I don't see how to check that.  How would I do this?

**Comment by jstedfast** [comment]
it's a `HashSet<string>` - it's just like any other collection of strings that you can enumerate.

**Comment by bender2k14** [comment]
I see. I thought you were saying that I could check if a given server supports a given mechanism.

So by default, `client.AuthenticationMechanisms` contains the mechanisms that MailKit supports, right?

**Comment by jstedfast** [comment]
No, client.AuthenticationMechanisms contains the list of mechanisms that the *server* supports.

**Comment by bender2k14** [comment]
Ok, good.  That is what I thought.  And per my report above, it is really the list of mechanisms that the server claims to support.

At one point the server administrator tried to justify their treatment of this mechanism by pointing it that it was last in the list of mechanisms that their server claims to support.

Given that `CRAM-MD5` is listed last, why does MailKit attempt to authenticate with it first?

**Comment by jstedfast** [comment]
The list of supported mechanisms has no bearing on preference according to the specs, so your admin has no idea what he's talking about.

CRAM-MD5 is used before LOGIN and PLAIN because LOGIN and PLAIN just base64 encode the password which is laughable security.

**Comment by bender2k14** [comment]
That makes sense.

Is there a way that I could have known that authentication via some mechanism was failing other than looking at the protocol log?  For example, is there an event that is raised each time some authentication mechanism fails?

**Comment by jstedfast** [comment]
Just the protocol log.

**Comment by bender2k14** [comment]
Ok, thanks.

Speaking of the protocol log, while I was investigating this problem, I modified a copy of [`ProcotolLogger`](https://github.com/jstedfast/MailKit/blob/master/MailKit/ProtocolLogger.cs) in order to log to [Serilog](https://serilog.net/).

In the process, I removed the "midline" functionality.  Said another way, each call to my implementation of `IProtocolLogger` corresponds to exactly one line in the corresponding log. In the cases that I analyzed, I still got the same output.  Maybe that logic isn't needed?

Anyway, here is my modification.  Maybe someone will find this useful.

```
using System;
using System.Text;
using Serilog;

namespace MailKit {
  /// <summary>
  /// Logs to Serilog.
  /// Intended to have the same line breaks as MailKit.ProtocolLogger.
  /// https://github.com/jstedfast/MailKit/blob/master/MailKit/ProtocolLogger.cs
  /// </summary>
  public class SerilogProtocolLogger : IProtocolLogger {

    private static readonly string _clientPrefix = "C: ";
    private static readonly string _serverPrefix = "S: ";

    private readonly ILogger _logger;

    public SerilogProtocolLogger(ILogger logger) {
      _logger = logger;
    }

    private static void ValidateArguments(byte[] buffer, int offset, int count) {
      if (buffer == null)
        throw new ArgumentNullException(nameof(buffer));

      if (offset < 0 || offset > buffer.Length)
        throw new ArgumentOutOfRangeException(nameof(offset));

      if (count < 0 || count > (buffer.Length - offset))
        throw new ArgumentOutOfRangeException(nameof(count));
    }

    private void _Log(string prefix, byte[] buffer, int offset, int count) {
      int endIndex = offset + count;
      int index = offset;
      int start;

      while (index < endIndex) {
        start = index;

        while (index < endIndex && buffer[index] != (byte) '\n') {
          index++;
        }

        if (index < endIndex && buffer[index] == (byte) '\n') {
          index++;
        }

        _logger.Verbose(prefix + Encoding.Default.GetString(buffer, start, index - start).Trim());
      }
    }

    public void LogConnect(Uri uri) =>
      _logger.Verbose("Connected to {URI}", uri);

    public void LogClient(byte[] buffer, int offset, int count) {
      ValidateArguments(buffer, offset, count);

      _Log(_clientPrefix, buffer, offset, count);
    }

    public void LogServer(byte[] buffer, int offset, int count) {
      ValidateArguments(buffer, offset, count);

      _Log(_serverPrefix, buffer, offset, count);
    }

    public void Dispose() { }

  }
}
```
